// composables/es/result.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _state = require('./state');

var _state2 = _interopRequireDefault(_state);

var _with = require('./with');

var _with2 = _interopRequireDefault(_with);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var initial = { value: undefined, pending: true, done: false, error: false };

var handlePromise = function handlePromise(promise, setResult) {
  var canceled = false;

  promise.then(function (value) {
    return !canceled && setResult({ value: value, pending: false, done: true, error: false });
  }).catch(function (value) {
    return !canceled && setResult({ value: value, pending: false, done: false, error: true });
  });

  return function () {
    return canceled = true;
  };
};

var Result = function Result(_ref) {
  var promise = _ref.promise,
      _render = _ref.render;
  return _react2.default.createElement(_state2.default, {
    initial: { result: initial },
    render: function render(_ref2) {
      var result = _ref2.result;
      return _react2.default.createElement(
        _react2.default.Fragment,
        null,
        _react2.default.createElement(_with2.default, {
          input: promise,
          enter: function enter(promise) {
            return handlePromise(promise, result.set);
          },
          exit: function exit(cancel) {
            return cancel() && result.set(initial);
          }
        }),
        _render(result.value)
      );
    }
  });
};

Result.propTypes = {
  promise: _propTypes2.default.object.isRequired,
  render: _propTypes2.default.func.isRequired
};

exports.default = Result;
